src/subtype.c

```diff
@@ -1779,7 +1781,10 @@ static jl_value_t *finish_unionall(jl_value_t *res JL_MAYBE_UNROOTED, jl_varbind
     jl_tvar_t *newvar = vb->var;
     JL_GC_PUSH2(&res, &newvar);
     // try to reduce var to a single value
-    if (obviously_egal(vb->lb, vb->ub)) {
+    if (jl_is_long(vb->ub) && jl_is_typevar(vb->lb)) {
```

Comment:
> jl_is_typevar(vb->lb)

IIUC, doesn't this also need checks that `vb->ub <: jl_any_type` && `vb->lb >: Union{}`?

Reply:
Yes, that would be more precise, I just wanted to be conservative here and also not complicate the code too much.